rfcs/20200422-concurrent-global-scratchpad-mode/README.md

```diff
@@ -0,0 +1,174 @@
+corresponding behavior discussed above, rather than serving as a hint.
+
+The following values for `dnnl_scratchpad_mode_t` will be introduced:
+
+```cpp
+/// The user enforces the concurrent mode scratchpad for the primitive
+/// implementation.
+dnnl_scratchpad_mode_concurrent,
+/// The user enforces the global mode scratchpad for the primitive
+/// implementation.
+dnnl_scratchpad_mode_global,
+```
+
+Similarly, the following `scratchpad_mode` enum values will be introduced:
+
+```cpp
+/// Scratchpad mode
+///
+concurrent = dnnl_scratchpad_mode_concurrent,
+///
+global = dnnl_scratchpad_mode_global,
+```
+
+## Compilation flag
+
+This proposal also aims to introduce a compilation flag
+`DNNL_DEFAULT_SCRATCHPAD_MODE`, which enables tweaking the scratchpad mode at
+compile-time for all primitives. It is proposed that this compilation flag will
+have the following three options:
+
+1. `DNNL_DEFAULT_SCRATCHPAD_MODE = DNNL_SCRATCHPAD_MODE_LIBRARY`. If specified,
+   the default scratchpad mode is equal to `scratchpad_mode::library`. This
+   maintain the current behavior of the library whereby each primitive decides
+   on its scratchpad mode.
+2. `DNNL_DEFAULT_SCRATCHPAD_MODE = DNNL_SCRATCHPAD_MODE_CONCURRENT`. If
+   specified, the default scratchpad mode is equal to
+   `scratchpad_mode::concurrent`, and enables a concurrent scratchpad for all
+   primitives.
+3. `DNNL_DEFAULT_SCRATCHPAD_MODE = DNNL_SCRATCHPAD_MODE_GLOBAL`. If specified,
+   the default scratchpad mode is equal to `scratchpad_mode::global`, and
+   enables a global scratchpad for all primitives.
+
+If `DNNL_DEFAULT_SCRATCHPAD_MODE` is not specified, the scratchpad mode in the
+library will default to `scratchpad_mode::library`.
+
+The above flags are expected to be beneficial in case a user does not require a
+per-primitive attribute and can rely on a global setting for scratchpad.
+Therefore, the users who do not use primitive attributes will benefit the most
+from the above compilation flags. However, the user who do use primitive
+attributes will not be impacted by usage of the above compilation flags since
+they only change the default `scratchpad_mode`, and it can be overriden by
+manually specifying a primitive attribute if necessary. 
+
+Given the above compilation flags, the proposal aims to replace
+`DNNL_ENABLE_CONCURRENT_EXEC` flag with `DNNL_DEFAULT_SCRATCHPAD_MODE =
+DNNL_SCRATCHPAD_MODE_CONCURRENT` since they are equivalent.
+
+## Usage models
+
+Since the new scratchpad modes are intended to complement existing one instead
+of replacing them, new usage models are assumed that may be favorable for the
+user in a particular instance. These usage models are briefly presented in the
+table below:
+
+| Scratchpad mode | Use case | Advantages |
+|-----------------|----------|------------|
+| **library**     | The library allocates scratchpad storage.<br>The mode is chosen by primitive implementation.                    | No API usage<br>     |
```

Comment:
What would be the benefit for a primitive to use private scratchpad (`concurrent`) vs a thread local memory pool 
 (`global`) by default?
And if no benefit, does it make sense to use `library` as a `library_default` mode, and it would be an alias to whatever the compilation flag specified for the default?

Reply:
The benefit of using `concurrent` is thread safety, enabling the user to create and execute one or more primitives on separate threads. However, if thread safety is not a concern and the user can create and execute primitives on the *same* thread, then the user is encouraged to use `global` scratchpad. The `global` mode is beneficial only in the context of a topology, where multiple primitives may create a scratchpad.

Regarding aliasing, I think `library_default` should remain independent of the compilation, since it emphasizes (at least to me) that the library/developer/primitive implementation decides on the scratchpad mode. Otherwise we may enforce the default to require a compilation flag (probably it's a question of preference).

Reply:
My question was about the `library` mode. I understand that a user might prefer thread safety (`concurrent`) over memory footprint (`global`).
What I do not understand is when the user does not care (`library` in your proposal), why would the library choose thread safety over memory footprint (and I understand even less that this decision is per primitive).

Reply:
I think initially it was all about thread safety (primitive private). Then, at some point, global scratchpad was introduced to reduce memory consumption. This led to the existing inconsistency since I don't think there are clear criteria over how to the library decides over scratchpad allocation.

Reply:
Yes, initially all scratchpads were private, and we added the global scratchpad to save memory footprint (and perf on systems with MCDRAM like Knights Landing).

Back to the initial point, 
1) I do not see when we would like to leave the memory management decision go to the primitives. 
2) we should have a default mode with the library wide policy specified by your proposed build option (needed for the C API).
So I was suggesting to just use your `library` mode as the `default` mode, and its behavior will be the specified by the build option.

Reply:
Indeed, the primitives will not decide how to manage scratchpad, here I actually meant the current behavior of the library where some primitives specify `GLOBAL` storage and other don't.

For your second point, I agree that the proposed build option should indicate what `scratchpad_mode::library` is. Also, the scope of this RFC covers CPU functionality only, and this mechanism will be documented along with the implementation. There I will also cover the rules for things like `get_scratchpad_mode` and `set_scratchpad_mode`, and how these can be impacted depending mostly on the engine.

Thanks a lot for your comments, I will update the RFC accordingly.

Reply:
I have updated the RFC, please review.